### PR TITLE
feat!: expedited proposals

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Feature: **2/3 Majority Emergency Voting Periods**
Original Spec: https://github.com/osmosis-labs/osmosis/issues/1517

#### Description

This change introduces an option for any existing proposal to be expedited. If a proposal is expedited, it uses a [shorter voting duration](https://github.com/osmosis-labs/cosmos-sdk/blob/de6d0d415d13880e37c980c100335a7b1d0d2808/x/gov/types/params.go#L15-L16) and a [higher tally threshold](https://github.com/osmosis-labs/cosmos-sdk/blob/de6d0d415d13880e37c980c100335a7b1d0d2808/x/gov/types/params.go#L23-L24). If a the higher threshold is not met under shorter voting duration, the proposal is converted as if it was a regular proposal and tallied / voted on under regular rules.

The distinction is achieved by having an [`IsExpedited`](https://github.com/osmosis-labs/cosmos-sdk/blob/de6d0d415d13880e37c980c100335a7b1d0d2808/proto/cosmos/gov/v1beta1/gov.proto#L53) flag on the proposal struct. When a proposal is converted from expedited to regular, this flag is flipped and [`VotingEndTime`](https://github.com/osmosis-labs/cosmos-sdk/blob/de6d0d415d13880e37c980c100335a7b1d0d2808/x/gov/types/gov.pb.go#L254) is extended. The rest of the fields,  including the `ProposalId` remain the same. As a result, all existing votes are preserved.

#### Differences from the Spec:

- `ExpeditedDuration` is implemented on the `VotingParams` as opposed to `Proposal` struct
- Did not create a separate index for expedited proposals. This would require duplicating logic for expedited proposals. Instead, in end block we iterate over both expedited and regular proposals and handle them appropriately

## Brief Changelog

- add `IsExpedited` flag to protos of all proposal content types
- add `ExpeditedVotingPeriod` to protos of gov voting parameters
- add `ExpeditedThreshold` to protos of gov tally parameters
- implement changes in `EndBlocker` of `x/gov` to move expedited proposals to regular (with extended voting period) once voting period ends with proposal not meeting `ExpeditedThreshold`
- add unit and integration tests to cover all new logic related to expedited proposals
- fix existing tests

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.

This change is already covered by existing tests, such as : 
- unit, integration and simulation tests in `x/gov`

This change added tests and can be verified as follows:
- `go test -timeout 30s -run ^TestExpeditedProposal_PassAndConversionToRegular$ github.com/cosmos/cosmos-sdk/x/gov`
- `go test -timeout 30s -run ^TestProposalPassedEndblocker$ github.com/cosmos/cosmos-sdk/x/gov`
- `go test -timeout 30s -run ^TestTickPassedVotingPeriod$ github.com/cosmos/cosmos-sdk/x/gov`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not documented

### CODE REVIEW

File: x/gov/abci.go [hunk 1]
```diff
@@ -49,12 +49,20 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
 
 		passes, burnDeposits, tallyResults := keeper.Tally(ctx, proposal)
 
-		if burnDeposits {
-			keeper.DeleteDeposits(ctx, proposal.ProposalId)
-		} else {
-			keeper.RefundDeposits(ctx, proposal.ProposalId)
+		// If an expedited proposal fails, we do not want to update
+		// the deposit at this point since the proposal is converted to regulat.
```
Comment (p0mvn) [inline x/gov/abci.go]:
```suggestion
		// the deposit at this point since the proposal is converted to regular.
```

File: x/gov/abci.go [hunk 2]
```diff
@@ -82,15 +90,32 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
...
 			}
 		} else {
-			proposal.Status = types.StatusRejected
-			tagValue = types.AttributeValueProposalRejected
-			logMsg = "rejected"
+			if proposal.GetContent().GetIsExpedited() {
+				// When expedited proposal fails, it is converted
+				// to a regular proposal. As a result, the voting period is extended and
+				// the tally is repeated according to the regular proposal rules,
+				//  once regular voting period expires again.
```
Comment (p0mvn) [inline x/gov/abci.go]:
```suggestion
				// to a regular proposal. As a result, the voting period is extended, and,
				// once the regular voting period expires again, the tally is repeated
				// according to the regular proposal rules.
```

File: x/gov/types/keys.go [hunk 3]
```diff
@@ -62,11 +63,22 @@ func GetProposalIDFromBytes(bz []byte) (proposalID uint64) {
...
+func ExpeditedProposalKey(proposalID uint64) []byte {
+	return append(ExpeditedProposalsKeyPrefix, GetProposalIDBytes(proposalID)...)
+}
+
+// GetProposalIdFromKey returns the proposalId given a prefixed byte key for expedited proposla.
+func GetProposalIdFromExpeditedKey(fullKey []byte) uint64 {
+	proposalIdBytes := fullKey[len(ExpeditedProposalsKeyPrefix):]
+	return GetProposalIDFromBytes(proposalIdBytes)
+}
+
```
Comment (p0mvn) [inline x/gov/types/keys.go]:
```suggestion
```

**Review by alexanderbez — COMMENTED** [review]
@p0mvn is there a reason why we define `is_expedited` on all proposal types (which are `Content` types), instead of directly on the `Proposal` type defined by `x/gov`?

**Comment by ValarDragon** [comment]
On chain dev call, we discussed moving this to the `Proposal` type, and breaking the CLI

**Comment by p0mvn** [comment]
The `isExpedited` field is now defined on the `Proposal` type

File: proto/cosmos/upgrade/v1beta1/upgrade.proto [hunk 4]
```diff
@@ -50,7 +50,7 @@ message SoftwareUpgradeProposal {
 
   string title       = 1;
   string description = 2;
-  Plan   plan        = 3 [(gogoproto.nullable) = false];
+  Plan   plan        = 4 [(gogoproto.nullable) = false];
```
Comment (p0mvn) [inline proto/cosmos/upgrade/v1beta1/upgrade.proto]: Just noticed - should change the numbers back. On it

File: proto/cosmos/distribution/v1beta1/distribution.proto [hunk 5]
```diff
@@ -110,8 +110,8 @@ message CommunityPoolSpendProposal {
 
   string   title                           = 1;
   string   description                     = 2;
-  string   recipient                       = 3;
-  repeated cosmos.base.v1beta1.Coin amount = 4
+  string   recipient                       = 4;
+  repeated cosmos.base.v1beta1.Coin amount = 5
```
Comment (mattverse) [inline proto/cosmos/distribution/v1beta1/distribution.proto]: What's the reason for changing the field numbers here?

File: proto/cosmos/params/v1beta1/params.proto [hunk 6]
```diff
@@ -13,7 +13,7 @@ message ParameterChangeProposal {
 
   string               title       = 1;
   string               description = 2;
-  repeated ParamChange changes     = 3 [(gogoproto.nullable) = false];
+  repeated ParamChange changes     = 4 [(gogoproto.nullable) = false];
```
Comment (mattverse) [inline proto/cosmos/params/v1beta1/params.proto]: same question here :)

File: proto/cosmos/distribution/v1beta1/distribution.proto (hunk 5 above)
Comment (alexanderbez) [inline proto/cosmos/distribution/v1beta1/distribution.proto]: I think you meant to undo these changes?

File: proto/cosmos/distribution/v1beta1/distribution.proto [hunk 7]
```diff
@@ -151,7 +151,7 @@ message CommunityPoolSpendProposalWithDeposit {
 
   string title       = 1 [(gogoproto.moretags) = "yaml:\"title\""];
   string description = 2 [(gogoproto.moretags) = "yaml:\"description\""];
-  string recipient   = 3 [(gogoproto.moretags) = "yaml:\"recipient\""];
-  string amount      = 4 [(gogoproto.moretags) = "yaml:\"amount\""];
-  string deposit     = 5 [(gogoproto.moretags) = "yaml:\"deposit\""];
+  string recipient   = 4 [(gogoproto.moretags) = "yaml:\"recipient\""];
+  string amount      = 5 [(gogoproto.moretags) = "yaml:\"amount\""];
+  string deposit     = 6 [(gogoproto.moretags) = "yaml:\"deposit\""];
```
Comment (alexanderbez) [inline proto/cosmos/distribution/v1beta1/distribution.proto]: ditto

File: proto/cosmos/gov/v1beta1/gov.proto [hunk 8]
```diff
@@ -170,9 +171,16 @@ message VotingParams {
     (gogoproto.jsontag)     = "voting_period,omitempty",
     (gogoproto.moretags)    = "yaml:\"voting_period\""
   ];
+  //  Length of the expedited voting period.
```
Comment (alexanderbez) [inline proto/cosmos/gov/v1beta1/gov.proto]:
```suggestion
  // Length of the expedited voting period.
```

File: proto/cosmos/gov/v1beta1/gov.proto [hunk 9]
```diff
@@ -192,9 +200,16 @@ message TallyParams {
...
+  bytes expedited_threshold = 3 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
+    (gogoproto.nullable)   = false,
+    (gogoproto.jsontag)    = "expedited_threshold,omitempty"
+  ];
+
   //  Minimum value of Veto votes to Total votes ratio for proposal to be
   //  vetoed. Default value: 1/3.
-  bytes veto_threshold = 3 [
+  bytes veto_threshold = 4 [
```
Comment (alexanderbez) [inline proto/cosmos/gov/v1beta1/gov.proto]: This should remain `3` and `expedited_threshold` should become `4`.

File: proto/cosmos/gov/v1beta1/gov.proto [hunk 10]
```diff
@@ -170,9 +171,16 @@ message VotingParams {
...
+  google.protobuf.Duration expedited_voting_period = 2 [
+    (gogoproto.nullable)    = false,
+    (gogoproto.stdduration) = true,
+    (gogoproto.jsontag)     = "expedited_voting_period,omitempty",
+    (gogoproto.moretags)    = "yaml:\"expedited_voting_period\""
+  ];
 
   // proposal_voting_periods defines custom voting periods for proposal types.
-  repeated ProposalVotingPeriod proposal_voting_periods = 2 [(gogoproto.nullable) = false];
+  repeated ProposalVotingPeriod proposal_voting_periods = 3 [(gogoproto.nullable) = false];
```
Comment (alexanderbez) [inline proto/cosmos/gov/v1beta1/gov.proto]: this should remain `2` and `expedited_voting_period` should be `3`.

File: proto/cosmos/params/v1beta1/params.proto (hunk 6 above)
Comment (alexanderbez) [inline proto/cosmos/params/v1beta1/params.proto]: undo?

File: proto/cosmos/upgrade/v1beta1/upgrade.proto (hunk 4 above)
Comment (alexanderbez) [inline proto/cosmos/upgrade/v1beta1/upgrade.proto]: undo?

**Review by alexanderbez — COMMENTED** [review]
So far I've just reviewed the proto changes, they look great apart from a few messed up field numbers (see comments). Will continue reviewing...

File: x/gov/abci.go [hunk 11]
```diff
@@ -49,12 +49,20 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
...
 		passes, burnDeposits, tallyResults := keeper.Tally(ctx, proposal)
 
-		if burnDeposits {
-			keeper.DeleteDeposits(ctx, proposal.ProposalId)
-		} else {
-			keeper.RefundDeposits(ctx, proposal.ProposalId)
+		// If an expedited proposal fails, we do not want to update
+		// the deposit at this point since the proposal is converted to regular.
+		// As a result, the deposits are either deleted or refunded in all casses
+		// BUT when an expedited proposal fails.
```
Comment (alexanderbez) [inline x/gov/abci.go]:
```suggestion
		// EXCEPT when an expedited proposal fails.
```

File: x/gov/types/genesis.go [hunk 12]
```diff
@@ -50,6 +50,18 @@ func ValidateGenesis(data *GenesisState) error {
 			threshold.String())
 	}
 
+	expeditedThreshold := data.TallyParams.ExpeditedThreshold
+	if expeditedThreshold.IsNegative() || expeditedThreshold.GT(sdk.OneDec()) {
+		return fmt.Errorf("governance expedited vote threshold should be positive and less or equal to one, is %s",
+			expeditedThreshold)
+	}
+
+	if expeditedThreshold.LT(threshold) {
```
Comment (alexanderbez) [inline x/gov/types/genesis.go]:
```suggestion
	if expeditedThreshold.LTE(threshold) {
```

File: x/gov/types/params.go [hunk 13]
```diff
@@ -152,7 +165,16 @@ func validateTallyParams(i interface{}) error {
...
-		return fmt.Errorf("vote threshold too large: %s", v)
+		return fmt.Errorf("vote threshold too large: %s", v.Threshold)
+	}
+	if !v.ExpeditedThreshold.IsPositive() {
+		return fmt.Errorf("expedited ote threshold must be positive: %s", v.ExpeditedThreshold)
+	}
+	if v.ExpeditedThreshold.GT(sdk.OneDec()) {
+		return fmt.Errorf("expedited vote threshold too large: %s", v.ExpeditedThreshold)
+	}
+	if v.ExpeditedThreshold.LT(v.Threshold) {
```
Comment (alexanderbez) [inline x/gov/types/params.go]:
```suggestion
	if v.ExpeditedThreshold.LTE(v.Threshold) {
```

**Review by alexanderbez — APPROVED** [review]
utACK. Looks great! Can't see any major flaws. Let's fix the proto stuff.

File: proto/cosmos/distribution/v1beta1/distribution.proto (hunk 5 above)
Comment (p0mvn) [reply proto/cosmos/distribution/v1beta1/distribution.proto]: Sorry, missed some of them. Will do a better job double-checking this time

File: x/gov/types/genesis.go (hunk 12 above)
Comment (p0mvn) [reply x/gov/types/genesis.go]: I will update locally because err messages need to be updated as well

File: x/gov/types/params.go (hunk 13 above)
Comment (p0mvn) [reply x/gov/types/params.go]: ditto: I will update locally because err messages need to be updated as well

**Review by alexanderbez — APPROVED** [review]
🚀

File: x/gov/keeper/proposal.go [hunk 14]
```diff
@@ -194,10 +194,10 @@ func (keeper Keeper) SetProposalID(ctx sdk.Context, proposalID uint64) {
 // voting period is used instead of the base voting period.
 func (keeper Keeper) ActivateVotingPeriod(ctx sdk.Context, proposal types.Proposal) {
 	proposal.VotingStartTime = ctx.BlockHeader().Time
-	proposal.Status = types.StatusVotingPeriod
 
-	votingPeriod := keeper.GetVotingPeriod(ctx, proposal.GetContent())
+	votingPeriod := keeper.GetVotingParams(ctx).GetVotingPeriod(proposal.IsExpedited)
 	proposal.VotingEndTime = proposal.VotingStartTime.Add(votingPeriod)
+	proposal.Status = types.StatusVotingPeriod
```
Comment (mattverse) [inline x/gov/keeper/proposal.go:200]: What's the reason for adding status voting period here?

File: x/gov/abci.go [hunk 15]
```diff
@@ -82,15 +90,30 @@ func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
...
+				// to a regular proposal. As a result, the voting period is extended, and,
+				// once the regular voting period expires again, the tally is repeated
+				// according to the regular proposal rules.
+				proposal.IsExpedited = false
+				votingParams := keeper.GetVotingParams(ctx)
+				proposal.VotingEndTime = proposal.VotingStartTime.Add(votingParams.VotingPeriod)
+
+				keeper.InsertActiveProposalQueue(ctx, proposal.ProposalId, proposal.VotingEndTime)
+
+				tagValue = types.AttributeValueProposalRejected
```
Comment (mattverse) [inline x/gov/abci.go]:
```suggestion
				tagValue = types.AttributeValueExpeditedProposalRejected
```

File: x/gov/types/keys.go [hunk 16]
```diff
@@ -42,6 +42,7 @@ var (
 	ActiveProposalQueuePrefix   = []byte{0x01}
 	InactiveProposalQueuePrefix = []byte{0x02}
 	ProposalIDKey               = []byte{0x03}
+	ExpeditedProposalsKeyPrefix = []byte{0x04}
```
Comment (mattverse) [inline x/gov/types/keys.go]: This key isn't being used I think it's good to delete this for now unless we have future plans to use this
```suggestion
```

File: x/gov/types/msgs_test.go [hunk 17]
```diff
@@ -27,26 +27,28 @@ func init() {
...
-		{"Test Proposal", "the purpose of this proposal is to test", ProposalTypeText, addrs[0], coinsMulti, true},
-		{strings.Repeat("#", MaxTitleLength*2), "the purpose of this proposal is to test", ProposalTypeText, addrs[0], coinsMulti, false},
-		{"Test Proposal", strings.Repeat("#", MaxDescriptionLength*2), ProposalTypeText, addrs[0], coinsMulti, false},
+		{"Test Proposal", "the purpose of this proposal is to test", true, ProposalTypeText, addrs[0], coinsPos, true},
+		{"", "the purpose of this proposal is to test", false, ProposalTypeText, addrs[0], coinsPos, false},
+		{"Test Proposal", "", true, ProposalTypeText, addrs[0], coinsPos, false},
+		{"Test Proposal", "the purpose of this proposal is to test", false, ProposalTypeText, sdk.AccAddress{}, coinsPos, false},
+		{"Test Proposal", "the purpose of this proposal is to test", true, ProposalTypeText, addrs[0], coinsZero, true},
+		{"Test Proposal", "the purpose of this proposal is to test", false, ProposalTypeText, addrs[0], coinsMulti, true},
+		{strings.Repeat("#", MaxTitleLength*2), "the purpose of this proposal is to test", true, ProposalTypeText, addrs[0], coinsMulti, false},
```
Comment (mattverse) [inline x/gov/types/msgs_test.go:42]: What do you think about adding both true & false for all possible test cases?

File: x/gov/types/proposal.go [hunk 18]
```diff
@@ -227,7 +232,7 @@ func RegisterProposalType(ty string) {
 }
 
 // ContentFromProposalType returns a Content object based on the proposal type.
-func ContentFromProposalType(title, desc, ty string) Content {
+func ContentFromProposalType(title, desc, ty string, isExpedited bool) Content {
```
Comment (mattverse) [inline x/gov/types/proposal.go]: `isExpedited` parameter isn't being used, could possibly delete this

File: x/gov/types/proposal.go [hunk 19]
```diff
@@ -13,14 +13,18 @@ import (
...
+const (
+	// DefaultStartingProposalID is 1
+	DefaultStartingProposalID uint64 = 1
+
+	errMsgFmtNotProtoMessage = "%T does not implement proto.Message"
+)
 
 // NewProposal creates a new Proposal instance
-func NewProposal(content Content, id uint64, submitTime, depositEndTime time.Time) (Proposal, error) {
+func NewProposal(content Content, id uint64, submitTime, depositEndTime time.Time, isExpedited bool) (Proposal, error) {
```
Comment (mattverse) [inline x/gov/types/proposal.go:24]: isExpedited parameter isn't being used, could possibly delete this

File: x/params/types/proposal/proposal.go [hunk 20]
```diff
@@ -22,7 +22,7 @@ func init() {
 	govtypes.RegisterProposalTypeCodec(&ParameterChangeProposal{}, "cosmos-sdk/ParameterChangeProposal")
 }
 
-func NewParameterChangeProposal(title, description string, changes []ParamChange) *ParameterChangeProposal {
+func NewParameterChangeProposal(title, description string, isExpedited bool, changes []ParamChange) *ParameterChangeProposal {
```
Comment (mattverse) [inline x/params/types/proposal/proposal.go:25]: isExpedited parameter isn't being used, could possibly delete this

**Review by mattverse — APPROVED** [review]
Nice and clean implementation 🔥 Just some minor nit comments and should be good to merge!

File: x/gov/abci.go (hunk 15 above)
Comment (alexanderbez) [reply x/gov/abci.go]: Good idea.

File: proto/cosmos/distribution/v1beta1/distribution.proto (hunk 5 above)
Comment (p0mvn) [reply proto/cosmos/distribution/v1beta1/distribution.proto]: I originally defined the `is_expedited` flag on each proposal separately. Then, changed it to be defined on the proposal struct for simplicity

When reverting this, I had some issues with saving protos and missed these. The numbering should be fixed now

File: proto/cosmos/params/v1beta1/params.proto (hunk 6 above)
Comment (p0mvn) [reply proto/cosmos/params/v1beta1/params.proto]: Also fixed

File: x/gov/types/keys.go (hunk 16 above)
Comment (p0mvn) [reply x/gov/types/keys.go]: Nice catch. Originally, was going to implement this via a second index but found a workaround

File: x/gov/types/msgs_test.go (hunk 17 above)
Comment (p0mvn) [reply x/gov/types/msgs_test.go:42]: There is no extra validation resulting from the "is expedited" flag at this time so this might be redundant. In my opinion, only if we add it in the future, does repeating all cases for true/false values would make sense.

File: x/gov/keeper/proposal.go (hunk 14 above)
Comment (p0mvn) [reply x/gov/keeper/proposal.go:200]: When we activate a proposal, it moves into the voting period state from [`StatusDepositPeriod`](https://github.com/osmosis-labs/cosmos-sdk/blob/a7d05e032ff30a35cea8301c908933ea210a6173/x/gov/types/gov.pb.go#L83)

Please note that this change is not new. Seems that I just changed the order of operations, and as a result, messing up the diff. Sorry if any confustion

File: x/gov/abci.go (hunk 15 above)
Comment (p0mvn) [reply x/gov/abci.go]: Nice catch

File: x/gov/types/msgs_test.go (hunk 17 above)
Comment (mattverse) [reply x/gov/types/msgs_test.go:42]: makes sense

**Comment by p0mvn** [comment]
Merging this to avoid having to continuously sync it with `osmosis-main`. We should be careful not to release this before governance decides on the future of this feature.

As for the current status of this as a whole, https://github.com/osmosis-labs/cosmos-sdk/pull/254 needs to be completed and merged before we create a gov prop. The gov prop is already written up and ready to go on-chain once #254 is in

**Comment by mattverse** [comment]
Sweet! lmk once #254  is r4r!
